### DISCUSSION TITLE
umi v4 不支持自定义 splitChunks 吗？

### DISCUSSION THREAD

**Discussion by Enivia** [umijs/umi#8513] [Q&A] [answered]
原项目中配置的 chunks 没有生效

**Comment by fz6m** [comment]
umi4 默认按页拆包，这近似等同于 umi3 中的 `dynamicImport` 。

对于特殊场景，可手动进一步拆包：

```ts
import { lazy, Suspense } from 'react'

// './Page' 该组件将被自动拆出去
const Page = lazy(() => import('./Page'))

export default function() {
  return (
    <Suspense fallback={<div>loading...</div}>
      <Page />
    </Suspense>
  )
}
```

**Reply by Enivia** [reply]
> umi4 默认按页拆包，这近似等同于 umi3 中的 `dynamicImport` 。
> 
> 对于特殊场景，可手动进一步拆包：
> 
> ```ts
> import { lazy, Suspense } from 'react'
> 
> // './Page' 该组件将被自动拆出去
> const Page = lazy(() => import('./Page'))
> 
> export default function() {
>   return (
>     <Suspense fallback={<div>loading...</div}>
>       <Page />
>     </Suspense>
>   )
> }
> ```

谢谢分享。但我想 split 的代码并不是组件级或页面级的，比如把所有 styles 打包成一份 css 文件，不借助 splitChunks 无法实现。

**Comment by Enivia** [comment, marked as answer]
找到了，可以这样自定义 cacheGroups。
```js
  chainWebpack(config) {
    config.optimization.splitChunks.merge({
      cacheGroups: {
        // ...
      },
    });
  },
```

文档需要补充额。

**Reply by lianer3588** [reply]
请问这样配置之后 还需要关闭 dynamicImport 吗？我配置了之后打包产物还是有按照分页打包的js

**Reply by fz6m** [reply]
umi4 最新版本已支持更多、更好的分包策略，无需手动修改 split chunk 配置，详见 [codeSplitting](https://umijs.org/docs/api/config#codesplitting)

**Reply by lianer3588** [reply]
codeSplitting配置之后 还是会按照页面去拆分js 如果项目页面较多，这样拆分太细了，js太多了，请问还有什么办法能减少打包产物吗？

**Reply by fz6m** [reply]
http 2 可以并发加载，越多并发加载越快，越好。

如果不需要的话，可以参考 [可以关闭 dynamicImport 吗？](https://umijs.org/docs/introduce/faq#%E5%8F%AF%E4%BB%A5%E5%85%B3%E9%97%AD-dynamicimport-%E5%90%97) 把 `.js` 产物合并成给一个。

**Reply by Enivia** [reply]
> 请问这样配置之后 还需要关闭 dynamicImport 吗？我配置了之后打包产物还是有按照分页打包的js

是否关闭主要看你自己的需要吧。dynamicImport 开启就是会按页面分离产物。

**Reply by destinytaoer** [reply]
为什么我用 umi4 打包出来，很多三方库都在 umi.js 文件里面，用 umi 的拆包策略完全没效果。umi.js 文件压缩过还有 2M+，有推荐的单独处理打包第三方库的方式吗

**Reply by fz6m** [reply]
自己配置下 [`codeSplitting`](https://umijs.org/docs/api/config#codesplitting) ，会自动拆包。

**Reply by destinytaoer** [reply]
三种自带的拆包策略都试过了，没啥用。公用的第三方库都打包进 umi.js 文件中了。umi 版本是 4.0.73。

**Reply by fz6m** [reply]
只有符合一定 size 大小的包才会被单独拆分，如果使用次数较多，则会被分配到公用 `umi.js` 产物中，可以使用 [`ANALYZE`](https://umijs.org/docs/guides/env-variables#analyze) 进行产物分析，另外对于特别大的组件和依赖部分，需要手动拆分，可参见 [代码拆分指南](https://umijs.org/blog/code-splitting) 。

如有问题，需要给一个最小复现。

**Reply by destinytaoer** [reply]
手动拆包只是拆分组件和页面级别的，实际上多页面和组件公共使用的部分还是会在 umi.js 中。我也尝试过的，只有单独被一个页面使用的三方库会被拆解出来，其他类似 antd 这种都会包含在 umi.js 里面。我是想问下有没有可以把 antd 这种大的、多页面和组件公用的三方库拆解出来的方式

**Reply by fz6m** [reply]
如果公用超过一定次数，就会被包含在 `umi.js` 中，umi 4 不支持配置分包配置，只能选择内置的 [`codeSplitting`](https://umijs.org/docs/api/config#codesplitting) 其中一种分包策略，如有特殊需求，可以把 antd 外置成 umd 格式的 script 作为 external 引入（仅生产）。

**Comment by HZZformGD** [comment]
...看了很久, 不知道什么时候能支持自定义分包设置, 目前的分包umi4 只能用给定的三种,   而打包后umi.js巨大(对于弱网环境来说),   想把一些不是首屏看到有关的js提取出来放到head里面,  尽可能减少首屏的加载方式, 不知道还有没有别的办法

**Reply by fz6m** [reply]
对于 `umi.js` 是公用的内容，无论如何哪个页面都需要加载这个基础库，其中包含了应用所需要的最基本依赖。

第一个因素是因为 react 应用本身就是 200kb （react + react-dom + react-router）起步的，所以 react 应用本来就是默认大的。

第二是近年 ecma 和浏览器支持的新语法快速增长， core-js 的 polyfill 越来越大，由于 umi 是一个面向中后台的框架，主要用于 PC 端，现在企业用户的网速都过剩，加上 cdn 分发、响应缓存、gzip 压缩，所以无需考虑加载缓慢的问题，况且一旦使用 图表、3d 相关的工具库立即就会让应用增长 几 M 、十几 M 产物，所以无需考虑应用产物体积问题，若面向的用户都是现代浏览器，可以通过配置较高的 [`targets`](https://umijs.org/docs/api/config#targets) 来减少 core-js 的 polyfill 含量。

另外还可以通过仅在生产环境 external 外置 react 、react-dom 的方式（ umd 外链 ）来单独提取 react 的依赖来减小单 js 的大小。

无论如何，react 应用在全部基础依赖准备完成之前都不会渲染，所以对于有足够网速的用户，除了 10M 以上的 js 文件，普通的资源分不分开 js 的区别不大。

当你需要考虑 toC 和手机端时，这不符合 umi 面向企业中后台用户的方向，若无法承担相关负面影响，则不应该使用 umijs 。

**Reply by 5201314999** [reply]
> 无论如何，react 应用在全部基础依赖准备完成之前都不会渲染，所以对于有足够网速的用户，除了 10M 以上的 js 文件，普通的资源分不分开 js 的区别不大。

网速有时侯波动 我们850k 的gzip 包还是会让用户感觉白屏很久

**Reply by 5201314999** [reply]
@fz6m 
![image](https://github.com/user-attachments/assets/0c69d9ad-c200-4716-8c13-39642c28549e)
![image](https://github.com/user-attachments/assets/4c389cae-3bf3-4ba9-846d-d9dfd9e81906)
这个时间用户就觉得受不了, umd 外链对版本感觉不太方便，升级的话
